src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java

```diff
@@ -0,0 +1,610 @@
+    }
+
+    /**
+     * Get the commit hash from the build map, the commit hash will be different for submission types or null.
+     *
+     * @param submissionType describes why the build was started.
+     * @return if the commit hash for the given submission type was found, otherwise null.
+     */
+    private Optional<String> getCommitHash(TestResults report, SubmissionType submissionType) {
+        final var assignmentSubmission = List.of(SubmissionType.MANUAL, SubmissionType.INSTRUCTOR).contains(submissionType);
+        final var testSubmission = submissionType == SubmissionType.TEST;
+        final var testSlugSuffix = RepositoryType.TESTS.getName();
+
+        // It's either an assignment submission, so then we return the hash of the assignment (*-exercise, *-studentId),
+        // or the hash of the test repository for test repo changes (*-tests)
+        return report.getCommits().stream().filter(
+                commit -> (assignmentSubmission && !commit.getRepositorySlug().endsWith(testSlugSuffix)) || (testSubmission && commit.getRepositorySlug().endsWith(testSlugSuffix)))
+                .map(Commit::getHash).findFirst();
+    }
+
+    @Override
+    public BuildStatus getBuildStatus(ProgrammingExerciseParticipation participation) {
+        final var isQueued = job(participation.getProgrammingExercise().getProjectKey(), participation.getBuildPlanId()).isInQueue();
+        if (isQueued) {
+            return BuildStatus.QUEUED;
+        }
+        final var projectKey = participation.getProgrammingExercise().getProjectKey();
+        final var planKey = participation.getBuildPlanId();
+        final var url = Endpoint.LAST_BUILD.buildEndpoint(JENKINS_SERVER_URL.toString(), projectKey, planKey).build(true).toString();
+        try {
+            final var jobStatus = restTemplate.getForObject(url, JsonNode.class);
+
+            return jobStatus.get("building").asBoolean() ? BuildStatus.BUILDING : BuildStatus.INACTIVE;
+        }
+        catch (HttpClientErrorException e) {
+            throw new JenkinsException("Error while trying to fetch build status from Jenkins for " + planKey, e);
+        }
+    }
+
+    @Override
+    public Boolean buildPlanIdIsValid(String projectKey, String buildPlanId) {
+        try {
+            jobXml(projectKey, buildPlanId);
+            return true;
+        }
+        catch (Exception emAll) {
+            return false;
+        }
+    }
+
+    @Override
+    public Optional<Result> retrieveLatestBuildResult(ProgrammingExerciseParticipation participation, ProgrammingSubmission submission) {
+        final var report = fetchLatestBuildResultFromJenkins(participation);
+
+        // The retrieved build result must match the commitHash of the provided submission.
+        if (report.getCommits().stream().map(Commit::getHash).noneMatch(hash -> hash.equals(submission.getCommitHash()))) {
+            return Optional.empty();
+        }
+
+        final var result = new Result();
```

Comment:
These lines are a duplicate of those in `createResultFromBuildResult`.

Reply:
True! Refactored it